src/txrequest.cpp

```diff
@@ -0,0 +1,722 @@
+    SequenceNumber m_current_sequence{0};
+
+    //! This tracker's priority computer.
+    const PriorityComputer m_computer;
+
+    //! This tracker's main data structure. See SanityCheck() for the invariants that apply to it.
+    Index m_index;
+
+    //! Map with this tracker's per-peer statistics.
+    std::unordered_map<NodeId, PeerInfo> m_peerinfo;
+
+    //! The current time. It is kept here internally to guarantee a view of time that does not go backwards.
+    std::chrono::microseconds m_now = std::chrono::microseconds::min();
+
+public:
+    void SanityCheck() const
+    {
+        // Recompute m_peerdata from m_index. This verifies the data in it as it should just be caching statistics
+        // on m_index. It also verifies the invariant that no PeerInfo announcements with m_total==0 exist.
+        assert(m_peerinfo == RecomputePeerInfo(m_index));
+
+        // Calculate per-txhash statistics from m_index, and validate invariants.
+        for (auto& item : ComputeTxHashInfo(m_index, m_computer)) {
+            TxHashInfo& info = item.second;
+
+            // Cannot have only COMPLETED peer (txhash should have been forgotten already)
+            assert(info.m_candidate_delayed + info.m_candidate_ready + info.m_candidate_best + info.m_requested > 0);
+
+            // Can have at most 1 CANDIDATE_BEST/REQUESTED peer
+            assert(info.m_candidate_best + info.m_requested <= 1);
+
+            // If there are any CANDIDATE_READY announcements, there must be exactly one CANDIDATE_BEST or REQUESTED
+            // announcement.
+            if (info.m_candidate_ready > 0) {
+                assert(info.m_candidate_best + info.m_requested == 1);
+            }
+
+            // If there is both a CANDIDATE_READY and a CANDIDATE_BEST announcement, the CANDIDATE_BEST one must be
+            // at least as good (equal or higher priority) as the best CANDIDATE_READY.
+            if (info.m_candidate_ready && info.m_candidate_best) {
+                assert(info.m_priority_candidate_best >= info.m_priority_best_candidate_ready);
+            }
+
+            // No txhash can have been announced by the same peer twice.
+            std::sort(info.m_peers.begin(), info.m_peers.end());
+            assert(std::adjacent_find(info.m_peers.begin(), info.m_peers.end()) == info.m_peers.end());
+        }
+    }
+
+    void PostGetRequestableSanityCheck() const
+    {
+        for (const Announcement& ann : m_index) {
+            if (ann.IsWaiting()) {
+                // REQUESTED and CANDIDATE_DELAYED must have a time in the future (they should have been converted
+                // to COMPLETED/CANDIDATE_READY respectively).
+                assert(ann.m_time > m_now);
+            } else if (ann.IsSelectable()) {
+                // CANDIDATE_READY and CANDIDATE_BEST cannot have a time in the future, as it would imply m_now
+                // going backwards.
+                assert(ann.m_time <= m_now);
+            }
```

Comment:
Just drop this now that m_time has no meaning for CANDIDATE_READY and CANDIDATE_BEST.

Reply:
Done.